Feature 288 memos translation to dynamodb

JuusoHeikkinen02 opened the pull request:
### Restore
Restored management memo endpoints for listing/fetching content from the google drive.

### Feature
Created a service for getting personal google authentication for accessing the drive (Service account will be implemented later so this will probably be removed).

Created a TranslatedMemoService for dynamoDB that has storeMemo and getTranslatedPdf functions.

Created handlers for creating/storing the translated memos and getting the translated memos from dynamoDB by id and targetLanguage.

### TODO
The translation implementation is not ready yet because we don't have the service for translation. There is commented code in the creation handler for the translation and I have tested it with mock data to ensure the sorting by targetLanguage works.

JuusoHeikkinen02 committed:
Created translated memo api service for creating and fetching translated memos from dynamodb.

Translation service is not yet implemented so the stored memos only have the original language.

Created handler for creating a new dynamodb memorecord from pdf fetched from the drive.

Modified the get translated memos handler to fetch from dynamodb with a primary key (memo id) and a sortkey (language fi/en)

currently in use is my personal google accounts authorization, will use service account in the future

JuusoHeikkinen02 committed:
Added spec reference

JuusoHeikkinen02 committed:
Added the spec model for the return body of the getTranslatedmemo.

Changed the endpoints from /google-drive to /translate-memo/...

Changed tsdocs to describe the params and returns

JuusoHeikkinen02 committed:
Merge remote-tracking branch 'origin/develop' into feature-288-memos-translation-to-dynamodb

JuusoHeikkinen02 committed:
removed unnecesary env from googleAuth

JuusoHeikkinen02 committed:
Removed all of the code related to the removed env

JuusoHeikkinen02 committed:
Modified the dynamodb service's tsdocs

fixed the functions index file's exports

JuusoHeikkinen02 commented on src/services/google-drive-api-service.ts:
```diff
@@ -210,60 +247,70 @@ export const getFolderId = async (year: string, month: string): Promise<string>
+      { responseType: "stream" }
+    );
     if (!response) {
       console.error(`Failed to fetch file content: ${response.status} - ${response.statusText}`);
```
removed unnecessary error catches, try catch should be fine for catching the errors

JuusoHeikkinen02 commented on src/services/google-drive-api-service.ts:
```diff
@@ -102,12 +124,12 @@ const getFolders = async (folderId: string) => {
 export const getBaseFolderByName = async (folderName: string) => {
   try {
     const folders = await getFolders(folderId);
-    const folderFound = folders.find(folder => folder.name === folderName);
+    const folderFound = folders.find((folder) => folder.name === folderName);
     return folderFound.id;
```
Added a descriptive error

JuusoHeikkinen02 committed:
Changed the capital letter of StoredMemo
Added error handling to the google-drive service

Jdallos commented on src/app/config.ts:
```diff
@@ -47,6 +51,12 @@ export default class Config {
       adminEmails: env.ADMIN_EMAILS ? env.ADMIN_EMAILS.split(",") : []
     },
+    google: {
+      clientId: env.GOOGLE_CLIENT_ID,
+      clientSecret: env.GOOGLE_CLIENT_SECRET,
+      refreshToken: env.GOOGLE_REFRESH_TOKEN,
+      driveFolderId: env.GOOGLE_DRIVE_FOLDER_ID
```
Are these likely to need changing when we get access to the workplace API? If so we should leave a comment

Jdallos commented on src/database/models/memo-record.ts:
```diff
@@ -0,0 +1,20 @@
+  translatedBase64: string;
+}
+
+export interface MemoInput {
```
Missing comment

Jdallos commented on src/database/services/translated-memo-api-service.ts:
```diff
@@ -0,0 +1,60 @@
+
+const TABLE_NAME = "TranslatedMemos";
+
+class TranslatedMemoService {
```
Comment?

Jdallos commented on src/database/services/translated-memo-api-service.ts:
```diff
@@ -0,0 +1,60 @@
+   *
+   * @returns The `MemoRecord` if found, or `null` if no matching memo exists.
+   */
+  public getTranslatedPdf = async (
```
Is it actually a Pdf we retrieve?

JuusoHeikkinen02 replied:
Well in this case we are retrieving the whole memorecord so might need to change the services name accordingly, also will make some changes to the handler so that we are sending the data in a actual pdf stream and not as JSON that contains the encoded base64 of the pdf.

Jdallos commented on src/functions/memo-management/drive-memos/translated-memos/create-translated-memo-pdf/index.ts:
```diff
@@ -0,0 +1,16 @@
+    {
+      httpApi: {
+        method: "POST",
+        path: "/translate-memo",
```
So this creates the translated memo, and expects to be provided a file id, but as this is a lambda function we are expecting to call it from the UI? So how will that work in terms of how does the user get the fileId (the google drive file) in the first place?
This might require some additional planning/ consulting the original thesis notes- do we list the files directly from the drive in the UI?

JuusoHeikkinen02 replied:
Yes the folders and files are listed first on the UI by year. Then the services fetch all folders/files by that year param so all of the needed fileIds will be accessible for the handler. One thing we need to consider is that when do we want to call this function exactly: when user selects the pdf from the drive to be displayed or only after user clicks the translation button on that pdf's page?

Jdallos commented on src/services/google-drive-api-service.ts:
```diff
@@ -20,59 +21,78 @@ export const getGoogleAuth = async () => {
+ *
+ * @returns Google Authentication
+ */
+export const getPersonalGoogleAuth = async () => {
```
So this will be redundant when we move to workspace API? IF so might be worth a comment

Jdallos commented on src/services/google-drive-api-service.ts:
```diff
@@ -116,27 +143,35 @@ export const getBaseFolderByName = async (folderName: string) => {
  * @returns array of File objects
  */
-export const getFilesInYear = async (year: string, mimeType: string = "application/pdf"): Promise<File[]> => {
+export const getFilesInYear = async (
```
So obviously the structure of these functions imply we expect some strucutre to the google drive? If so it could be worth documenting this- perhaps in Alex's onboarding docs but also in the api spec in the related endpoint

Jdallos commented on src/services/google-drive-api-service.ts:
```diff
@@ -210,60 +252,66 @@ export const getFolderId = async (year: string, month: string): Promise<string>
 /**
  * Creates a new PDF file from binary content and stores it in a specified folder
- * 
+ *
```
We should probably remove this createPdfFile and all references as I'm pretty sure we don't want to be creating anything in the drive- read only

JuusoHeikkinen02 committed:
Added todo comments for the google envs and the personalGoogleAuth.

Made some changes to the service names so they represent the logic better.

Removed the function to create a Pdf because its not needed.

Changed the return data of the get translated memo pdf to return the actual pdf stream when the handler is called (previously it returned the data as an JSON body containing all of the data of the memoRecord)

JuusoHeikkinen02 committed:
Add TranslatedMemos table to DynamoDB configuration

JuusoHeikkinen02 committed:
Merge remote-tracking branch 'origin/develop' into feature-288-memos-translation-to-dynamodb

JuusoHeikkinen02 committed:
Spec reference update

JuusoHeikkinen02 committed:
updated tsdocs
